Libraries/SDHC/ff13/Source/diskio.c

```diff
@@ -31,23 +33,44 @@ static unsigned int init_done = 0;
 
 /*local vaiables*/
 static uint8_t rtc_en;
-
+uint8_t flash = 0;
```

Comment:
I am a little bit confused about using the global `flash` variable vs `#ifdef FLASH` in the preprocessor.

Consider the following function:

```C
DSTATUS disk_status (
    BYTE pdrv       /* Physical drive nmuber to identify the drive */
)
{
	if(!flash){
    	DSTATUS status = 0;

    	if (!MXC_SDHC_Card_Inserted()) {
			init_done = 0;
			status = STA_NOINIT | STA_NODISK;
    	}

    	return status;
	}
#ifdef FLASH
	else{
		    DSTATUS status = 0;


    if (mscmem_ID() == MX25_EXP_ID) {
    	status = 0;
    }
    else{
    	init_done = 0;
    	status = STA_NOINIT | STA_NODISK;
    }

    return status;
	}
#endif
	return 0;
}
```

Why is the global runtime variable necessary?  I don't see any calls to `disk_flash` in this driver code.  Given the compiler definition and the suggestion above we can reduce this function to:

```C
DSTATUS disk_status (
    BYTE pdrv       /* Physical drive nmuber to identify the drive */
)
{
    DSTATUS status = 0;
#ifndef FLASH
    init_done = MXC_SDHC_Card_Inserted();
#else
    init_done = mscmem_ID();
#endif
    if (!init_done) {
        status = STA_NOINIT | STA_NODISK;
    }
	return status;
}
```

Reply:
The global external_flash variable is necessary because the new AppSwitcher code will check whether the user uses an SD Card or the external flash (if the SD Card is not inserted, then continue with the external flash). I added a preprocessor because if an application other than AppSwitcher calls the diskio.c, it is not necessary to compile all mscmem functions. The disk_flash() is called from the main.c of the new AppSwitcher, which also exists as a PR in another repository (refdes).

Reply:
Ok, thank you for clarifying.  This situation is a multi-drive system:

![image](https://user-images.githubusercontent.com/38844790/236076001-f02933f6-3c9c-496b-ba05-94dbe93bed2b.png)
(http://elm-chan.org/fsw/ff/doc/appnote.html#port)

The [disk_initialize](http://elm-chan.org/fsw/ff/doc/dinit.html) function is already designed to handle this with the `pdrv` argument.  In the image above physical drive 2 is the SD card slot, and physical drive 0 is the external flash.

Following that scheme, the app switcher application code would first attempt to initialize drive 2 (SD).  If it fails, then you attempt to initialize drive 0 (Ext Flash).

The `diskio.c` file in the avr example in [ffsample.zip](https://github.com/Analog-Devices-MSDK/msdk/files/11392032/ffsample.zip) is a decent reference for a multi-drive implementation.

Notice how they are separating the driver layers based on a scheme they have defined for `pdrv`.

```C
/*-----------------------------------------------------------------------*/
/* Low level disk I/O module glue functions         (C)ChaN, 2016        */
/*-----------------------------------------------------------------------*/
/* If a working storage control module is available, it should be        */
/* attached to the FatFs via a glue function rather than modifying it.   */
/* This is an example of glue functions to attach various exsisting      */
/* storage control modules to the FatFs module with a defined API.       */
/*-----------------------------------------------------------------------*/

#include "ff.h"			/* Obtains integer types for FatFs */
#include "diskio.h"		/* FatFs lower layer API */
#ifdef DRV_CFC
#include "cfc_avr.h"	/* Header file of existing CF control module */
#endif
#ifdef DRV_MMC
#include "mmc_avr.h"	/* Header file of existing SD control module */
#endif

// ...

/*-----------------------------------------------------------------------*/
/* Inidialize a Drive                                                    */
/*-----------------------------------------------------------------------*/

DSTATUS disk_initialize (
	BYTE pdrv				/* Physical drive nmuber to identify the drive */
)
{
	switch (pdrv) {
#ifdef DRV_CFC
	case DRV_CFC :
		return cf_disk_initialize();
#endif
#ifdef DRV_MMC
	case DRV_MMC :
		return mmc_disk_initialize();
#endif
	}
	return STA_NOINIT;
}
```